### ISSUE TITLE
Batching-isolation check fails on GLM Flash hybrid with default dense kernels

### ISSUE THREAD

**Issue by HawkBearPig** [HawkBearPig/dgpp#32]
The current `serve_load.py --isolation` check reproducibly fails for GLM-5.3-Flash NVFP4/FP8 with the default dense-kernel configuration. A controlled C1–C4 sweep points to numerical batch sensitivity: increasing `DGPP_DENSE_GEMV_ROWS` restores exact text agreement within each tested configuration. This report does **not** establish request-state contamination.

## Environment

- Source: `a3ed8994a34cf05f79e63dc1ad51ffd1f65c9f82`; measured 2026-09-22 UTC.
- Release server: `0.1.0+ga3ed8994a34c`; SHA-256 `bdf69fe085e06b757ee26fb4d236ab3fafd71378f8ead88685209723d46e607c`.
- Four NVIDIA GB10 DGX Spark nodes, 121 GiB unified memory per node, 200 Gb/s RoCE; CUDA 13.0, driver 580.173.02.
- Model: `HawkBearPig/GLM-5.3-Flash-NVFP4-FP8`.
- Configuration: `deploy/cluster_glm-5.3-flash_nvfp4-fp8_w4.example.json`, four slots, default MTP, prefix cache enabled.
- Temperature 0, maximum 256 output tokens, server-default thinking. Rank 0 uses `CUDA_DEVICE_MAX_CONNECTIONS=32`; peers inherit their login environment. `DGPP_DENSE_GEMV_ROWS` is propagated to all ranks by the launcher.

## Reproduction

With prepared checkpoints and the cluster site configured, build the revision above and run on an otherwise idle cluster:

```bash
cmake --preset release
cmake --build --preset release -j 4
export CUDA_DEVICE_MAX_CONNECTIONS=32
export DGPP_DENSE_GEMV_ROWS=4
CONFIG=deploy/cluster_glm-5.3-flash_nvfp4-fp8_w4.example.json
python3 scripts/dgpp-cluster up --config "$CONFIG" --bin build-release/dgpp-serve
python3 scripts/serve_load.py 127.0.0.1 18080 \
  --concurrency 1 --max-tokens 256 --isolation 4 --classes chat \
  --json-out /tmp/isolation.json
python3 scripts/dgpp-cluster down --config "$CONFIG"
```

The isolation check uses `PROMPTS[0]` from the mixed corpus, independently of `--classes chat`. GLM-5.3 rejects the no-thinking toggle, so the client retries with the server-default thinking mode.

Observed output:

```text
== isolation: prompt 0 alone (256 tokens) vs beside 3 others (256 tokens): DIFFERENT
   first difference at char 495: alone 'poor sharing\n3. Paged KV cache mechanics: blocks, block size' | batched 'inflexible scheduling\n3. Paged KV cache mechanics: blocks, b'
transcript isolation failed
```

The two-node hybrid deployment also failed its C4 exact-text check (first difference at character 338). The kernel-setting controls below were performed on the four-node deployment only.

## Controlled results

Each setting used a fresh server with the same binary. After one 64-token warmup, the same probe was run at concurrency `1, 1, 2, 2, 3, 3, 4, 4, 1, 1`, with distinct companion prompts. Full responses, usage and reconciled scheduler counters were captured. Each result was compared with the first solo response from that configuration.

| `DGPP_DENSE_GEMV_ROWS` | C1 | C2 | C3 | C4 | Return to C1 |
|---|---|---|---|---|---|
| 4 (default) | Identical | Identical | Different | Different | Identical |
| 8 | Identical | Identical | Identical | Identical | Identical |
| 256 | Identical | Identical | Identical | Identical | Identical |

Each cell covers two repetitions. At the default setting, all four C3/C4 results diverged at character 495 and had the same response hash. All probe requests reported 62 prompt tokens, including 60 cached tokens. All four rank operation-stream hashes matched in each diagnostic run; this checks execution order, not numerical equality or request isolation.

Response SHA-256 values:

| Setting | C1/C2 | C3/C4 |
|---|---|---|
| 4 | `8d354804060ca574b6ff81fbb4a6cca9c22793fa7614bdbcd7ebda6ba83a6ca4` | `11f99abc53960bc633aeec53a688165d60d6ae851bff61f5c67cc0f76818fe3b` |
| 8 | `338bc3fc23a42c709ddd9e493e47e64b258cae9fee56f9d56b6bc429b79da7d3` | same |
| 256 | `37225350895f9b5f9fb29991072ec041be2064de1e9d6a57cf34449f59748f9f` | same |

Changing the kernel setting also changes the solo transcript. The controls establish agreement **within** each setting, not equality between settings. Diagnostic timings are excluded from the benchmark performance tables.

## Interpretation and follow-up

The [dense-kernel contract](https://github.com/HawkBearPig/dgpp/blob/a3ed8994a34cf05f79e63dc1ad51ffd1f65c9f82/src/kernels/gemm.hpp#L83-L96) explicitly permits different reduction orders across batch shapes. The [engine test](https://github.com/HawkBearPig/dgpp/blob/a3ed8994a34cf05f79e63dc1ad51ffd1f65c9f82/tests/cuda/glm_dsa_engine_test.cpp#L168-L178) allows decision changes around near ties. In contrast, [the service check](https://github.com/HawkBearPig/dgpp/blob/a3ed8994a34cf05f79e63dc1ad51ffd1f65c9f82/scripts/serve_load.py#L314-L340) demands exact transcript equality and labels a mismatch an isolation failure.

The controlled results support kernel dispatch as the explanation for this particular mismatch. We have not captured logits at the first divergence, established an error bound, or ruled out every possible state-isolation defect.

Suggested resolution:

1. Define the supported batching guarantee and make the service check/documentation consistent with it. Distinguish exact greedy reproducibility from request-state isolation.
2. Add a meaningful correctness check for the default batched kernels: compare logits at a shared prefix and quantify the first divergent decision, alongside request-state isolation coverage. Do not silently accept arbitrary text differences as harmless.
3. Preserve both responses and request settings when the exact-text check fails. Currently it exits before saving the requested JSON, leaving only a short excerpt.

The benchmark refresh retains the failing exact-text result and the diagnostic evidence; it does not change kernel defaults to obtain a passing result.

## Diagnostic client

Save the following as `/tmp/isolation_probe.py`, run it from the repository root against each freshly started server, and use a distinct `--json-out` path for each setting. It uses only modules already present at the recorded revision.

```python
#!/usr/bin/env python3
"""Capture solo/batched greedy responses and counters for the isolation prompt."""
import argparse
import hashlib
import json
from pathlib import Path
import sys

ROOT = Path.cwd()
sys.path.insert(0, str(ROOT / "scripts"))
from serve_load import PROMPTS, served_model, stream_one
from serve_prefill_probe import get
from timed_load import TimedLoad

parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument("--json-out", type=Path, required=True)
args = parser.parse_args()
host, port = "127.0.0.1", 18080
model = served_model(host, port)
load = TimedLoad()
load.settle(host, port, [])
warm = {}
stream_one(host, port, model, PROMPTS[0], 64, True, warm)
if warm.get("error"):
    raise RuntimeError(warm["error"])
load.settle(host, port, [warm])
report = {"model": model, "temperature": 0, "max_tokens": 256,
          "thinking": "server_default", "warmup": warm, "phases": []}
for c in (1, 1, 2, 2, 3, 3, 4, 4, 1, 1):
    before = get(host, port, "/v1/metrics")
    load.phase(host, port, model, c, 256, True, 0, PROMPTS, 0,
               f"isolation-c{c}", report["phases"])
    phase = report["phases"][-1]
    phase["metrics_before"] = before
    phase["metrics_after"] = get(host, port, "/v1/metrics")
    answer = phase["requests"][0]
    baseline = report["phases"][0]["requests"][0]
    a, b = baseline["text"], answer["text"]
    first = next((i for i, (left, right) in enumerate(zip(a, b)) if left != right), min(len(a), len(b)))
    phase["probe"] = {"text_sha256": hashlib.sha256(b.encode()).hexdigest(),
                      "identical_to_first_solo": (a, baseline["tokens"]) == (b, answer["tokens"]),
                      "first_difference": None if a == b else first,
                      "solo_excerpt": a[first:first + 120], "batch_excerpt": b[first:first + 120]}
    args.json_out.write_text(json.dumps(report, indent=2, allow_nan=False) + "\n")
    print("PROBE", c, json.dumps(phase["probe"]), flush=True)
```

```bash
python3 /tmp/isolation_probe.py --json-out /tmp/isolation-default.json
```
